renku/core/incubation/database.py

```diff
@@ -0,0 +1,470 @@
+            raise errors.ObjectStateError(f"Object '{object._p_oid}' already has a Database: '{object._p_jar}'")
+
+    def get(self, oid: Union[bytes, str]) -> Persistent:
+        """Get the object by oid."""
+        if oid in self._root_object_types:
+            return self.root[oid]
+
+        if isinstance(oid, str):
+            oid = oid.encode("utf-8")
+
+        object = self._cache.get(oid)
+        if object is not None:
+            return object
+        object = self._added.get(oid)
+        if object is not None:
+            return object
+        object = self._pre_cache.get(oid)
+        if object is not None:
+            return object
+
+        data = self._storage.load(oid)
+        object = self._reader.deserialize(data)
+        object._p_changed = 0
+
+        # NOTE: Avoid infinite loop if object tries to load its state before it is added to the cache
+        self._pre_cache[oid] = object
+        self._cache[oid] = object
+        self._pre_cache.pop(oid)
+
+        return object
+
+    def commit(self):
+        """Commit modified and new objects."""
+        for object in self._registered_objects:
+            oid = object._p_oid
+            assert oid
+
+            if object._p_jar is not self:
+                raise errors.ObjectStateError(f"Object '{object}' doesn't belong to DataManager.")
+            elif not object._p_changed and object._p_serial != NEW:
+                continue
+
+            object_type = type(object).__name__
+            if object_type in self._root_object_types:
+                self._create_root_object_type(self.root, object_type)
+                oid = oid.decode("ascii")
+                self.root[object_type][oid] = object
+
+            self._store_object(object)
+
+    def _store_object(self, object: Persistent):
+        writer = ObjectWriter(object)
+
+        for object in writer:
+            oid = object._p_oid
+
+            object._p_jar = self
+            data = writer.serialize(object)
+
+            self._storage.store(oid=oid, data=data)
+            self._cache[oid] = object
+
+            self._cache.update_object_size_estimation(oid, 1)
+            object._p_estimated_size = 1
+
+            object._p_changed = 0  # NOTE: transition from changed to up-to-date
+            object._p_serial = PERSISTED
+
+    def oldstate(self, object, tid):
+        """See persistent.interfaces.IPersistentDataManager::oldstate."""
+        raise NotImplementedError
+
+    def setstate(self, object):
+        """Load the state for a ghost object."""
+        oid = object._p_oid
+
+        data = self._storage.load(oid)
+        self._reader.set_ghost_state(object, data)
+
+        object._p_serial = PERSISTED
+
+    def register(self, object):
+        """See persistent.interfaces.IPersistentDataManager::register."""
+        self._registered_objects.append(object)
+
+    @classmethod
+    def from_path(cls, path: Union[str, Path]) -> "Database":
```

Comment:
Actually, `Database` used to get a `Storage` in its constructor which I've changed it to create the storage itself. On a second thought, that approach was cleaner in case we want to create a test database with a test storage.